### PR TITLE
[FLINK-27159][table-api] Support first_value/last_value in the Table API

### PR DESCRIPTION
**Pull request by shuiqiangchen**
## What is the purpose of the change

This pull request added support for `first_value` and `last_value` in Table API.

## Brief change log

- Added `first_value` and `last_value` in BuiltInFunctionDefinitions
- Added `first_value` and `last_value` in SqlAggFunctionVisitor

## Verifying this change

This change has test case covered by testFirstLastValue() in AggregateITCase.scala

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not documented)

### CODE REVIEW

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java [hunk 1]
```diff
@@ -410,6 +410,20 @@ ANY, and(logical(LogicalTypeRoot.BOOLEAN), LITERAL)
                     .outputTypeStrategy(TypeStrategies.aggArg0(t -> t, true))
                     .build();
 
+    public static final BuiltInFunctionDefinition FIRST_VALUE =
```
Comment (dianfu) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java:413]: Add last_value/first_value methods in BaseExpressions and also in PyFlink?

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/runtime/stream/table/AggregateITCase.scala [hunk 2]
```diff
@@ -114,6 +114,34 @@ class AggregateITCase(mode: StateBackendMode) extends StreamingWithStateTestBase
     assertEquals(expected.sorted, sink.getRetractResults.sorted)
   }
 
+  @Test
+  def testFirstLastValue(): Unit = {
```
Comment (dianfu) [inline flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/runtime/stream/table/AggregateITCase.scala:118]: Could we integrate it into existing test case, e.g. testGroupAggregateWithExpression, etc. Then we could avoid introducing another it case.

**Review by dianfu — COMMENTED** [review]
@shuiqiangchen Thanks for the PR. Have left a few minor comments. Besides, it would be great to also update the documentation: https://nightlies.apache.org/flink/flink-docs-release-1.15/docs/dev/table/functions/systemfunctions/.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java (hunk 1 above)
Comment (shuiqiangchen) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java:413]: Ok, It was my omission

**Comment by dianfu** [comment]
@shuiqiangchen It seems that there is a new way provided to support built-in functions. See https://issues.apache.org/jira/browse/FLINK-20522 and https://github.com/apache/flink/commit/df46f8948279d1ed89d7113919b4c4f720cf4199 for more details.

**Comment by shuiqiangchen** [comment]
@dianfu Thanks, I will refer to those commits in the ticket you mentioned.

**Comment by dianfu** [comment]
@shuiqiangchen It seems that the new stack is used when adding new functions. It makes adding new functions easier. Regarding to this PR, its purpose is to expose existing functions first_value/last_value to Table API and so I guess we could still use the old stack.

**Comment by shuiqiangchen** [comment]
@dianfu Yes, the implementation of LastValueAggFunction and FirstValueAggFunction have followed the new form. This PR is mainly to expose builtin functions to TableAPI.

**Comment by shuiqiangchen** [comment]
@dianfu Thank you for your suggestions. I have updated the PR accordingly, please have a look.
Changes are as followed:
- added firstValue and lastValue in BaseExpressions
- added first_value and last_value in expression.py.
- Reused the existing ITCase.
- updated the documentation for System (Built-in) Functions

And there is a minor issue , I find that the functions documentation for python is redirect to System (Built-in) Functions documentation for Java API.  It's first_value() in python while firstValue() in Java, which seems unaligned that the doc might be a bit misleading.

**Comment by dianfu** [comment]
> And there is a minor issue , I find that the functions documentation for python is redirect to System (Built-in) Functions documentation for Java API. It's first_value() in python while firstValue() in Java, which seems unaligned that the doc might be a bit misleading.
Good catch. We could find a way to improve this in a separate ticket.

**Review by dianfu — APPROVED** [review]

**Pull request closed without merging**